### ISSUE TITLE
Semantics of YR_UNDEFINED since 4.x

### ISSUE THREAD

**Issue by metthal** [VirusTotal/yara#1385]
There was a change in YARA 4.x which changed a semantics of `YR_UNDEFINED` in certain operations which are `==`, `!=`, `<`, `<=`, `>`, `>=`, `contains` and `matches`. Basically whenever one of the operand is undefined, it no longer evaluates to undefined but rather to `false`. This means that these 2 expressions

```yara
not pe.signatures[0].subject == "..."
pe.signatures[0].subject != "..."
```

will evaluate differently if there are no signatures in the scanned file. First one will be `true` (`==` evaluates to `false`, `not` makes it `true`) while the other one will be `false` (`!=` evaluates to `false`). In 3.x, neither of them will evaluate to `true`.

My question is, is this intended behavior? I suppose you should always check for these undefined cases but in certain scenarios, there's no way (or at least very limited way) to check for them like in case of `pe.imphash()`, `pe.rich_signature` etc. Unless there is something like `is_def(x)` function in YARA, there won't be a proper way to check these values. This all gives me an impression that this might be undesired behavior and that's why I am asking.

**Comment by wxsBSD** [comment]
This is somewhat discussed in https://github.com/VirusTotal/yara/issues/1238. I don't think it's undesired behavior, just a change in behavior. 😐

**Comment by metthal** [comment]
Thanks @wxsBSD for linking this issue. I welcome any change in this regard since propagating `UNDEFINED` even through logical operators was a bad idea in 3.x as it will render a lot of conditions unusable. However, I don't think #1180 was fixed correctly. `not` operator returning `false` from `UNDEFINED` is fine, but since also relational operators evaluate to `false`, expressions like the one in my example are evaluated in a very strange way.

I don't think it's correct that 2 equal expressions are evaluated in a different way (once as `true` and other time as `false`). This behavior only introduces confusion. Let's say you want to express that certain signature does not equal to certain string. If you write it as `pe.signatures[0].subject != "..."` then you'll end up with `false` when there are no signatures. But if you write it as `not pe.signatures[0].subject == "..."` then you'll end up with `true` if there are no signatures I've seen both styles being used in real world rules. I don't really think this is something you should keep in mind when writing rules.

**Comment by plusvic** [comment]
I  totally agree, but we have a problem here because we can't easily solve both this and #1180 at the same time. If we make operations `==`, `!=`, `contains`, etc to return `UNDEFINED` when some of argument is undefined, this issue can be fixed but #1180 gets broken again.

The only solution I see is removing the concept of `UNDEFINED` altogether and returning zero values for things that are undefined, for example `pe.signatures[0].subject` would be an empty string and `pe.entry_point` would be zero. This has some implications, because conditions like `pe.entry_point == 0` that now are false for non-PE files, will become true with this new logic.

But I don't see a way of solving all the edge cases of using `UNDEFINED`, specially when the `not` operator is involved.

**Comment by plusvic** [comment]
Mmmm... maybe we can do something if some operators return `false` when an argument is undefined and while others return `true` in the same case. I'm going to think carefully about it.

**Comment by metthal** [comment]
Yeah, I have been thinking about it these 2 days but I still haven't come up with anything that'd work in all cases. I am thinking about it too and will post here once I figure something out.

**Comment by plusvic** [comment]
I've thinking about this problem and didn't found an elegant solution. At first I thought about making the `==` operator return `false` if any the operands was undefined, and `!=` return `true` if any of the operands was undefined. This apparently solves our problems, but what can we do with operators `>`, `<`, `>=`, and `<=`? 

The result from `undefined < "foo"` should be the same than `not undefined >= "foo"`, but in that case  `undefined >= "foo"` should be the opposite of `undefined < "foo"`. These relationship operators introduce a lot of problems.

At the end I'm thinking about reverting b7da5d2835cc4cf8a15027da30265addab1a4be5 and go back to the old semantics. This reintroduces #1180, but I think it's the most coherent solution. Thoughts?
